README.md

```diff
@@ -103,7 +103,7 @@ Options
 |   | [`CIBW_BUILD`](https://cibuildwheel.readthedocs.io/en/stable/options/#build-skip)  <br> [`CIBW_SKIP`](https://cibuildwheel.readthedocs.io/en/stable/options/#build-skip)  | Choose the Python versions to build |
 | **Build customization** | [`CIBW_ENVIRONMENT`](https://cibuildwheel.readthedocs.io/en/stable/options/#environment)  | Set environment variables needed during the build |
 |   | [`CIBW_BEFORE_ALL`](https://cibuildwheel.readthedocs.io/en/stable/options/#before-all)  | Execute a shell command on the build system before any wheels are built. |
-|   | [`CIBW_BEFORE_BUILD`](https://cibuildwheel.readthedocs.io/en/stable/options/#before-build)  | Execute a shell command preparing each wheel's build |
+|   | [`CIBW_BEFORE_BUILD`](https://cibuildwheel.readthedocs.io/en/stable/options/#before-build)  | Execute a shell command preparing each wheel's build if pyproject.toml is not enough |
```

Comment:
Okay, but the reason I was heavy handed here is the problem is entirely cibuildwheel's fault. If your project supports `pip install .`, it supports `pip wheel .`; as long as wheel is installed (it is), pip builds a wheel first before installing. In fact, that's entirely the problem that causes cibuildwheel to have this workaround. `pip wheel .` says "build all wheels required to install this project", which is exactly what `pip install` needs, but not what a package builder needs, since you don't want to build dependent projects wheel's (looking at you, NumPy on weird platforms or very new Pythons...). So we implement a hack - we add `--no-deps`. Now there's an asymmetry between `pip install .` and `pip wheel . --no-deps`, and *that's* why this is here.

I'd like to do a proper study of this someday, and see if we can improve the workaround. Moving to pypa-build might be the correct solution, but there might be a better way to solve it as it exists.

Reply:
Wait, I'm not sure I'm getting this; `--no-deps` disables `install_requires` and `setup_requires`, right? `install_requires` is not an issue, because we do allow dependencies when installing before testing, and I thought `setup_requires` was deprecated/very fickle to get working/not advised, because it's using the old eggs and not doing a proper install anyway)?

But something tells me I'm missing something here?

Reply:
Yes, this is exactly right. Users should not be using setup_requires, but should be using pyproject.toml. But we are discussing the workaround, `CIBW_BEFORE_BUILD: pip install .`, which (see #523 ) is _only_ useful if you have a package in `setup_requires` and `install_requires`.

Reply:
Right, yes. `pip install .` in `CIBW_BEFORE_BUILD` is a horrible workaround, indeed. But `CIBW_BEFORE_BUILD` can also handle non-pip dependencies, ofc.

Reply:
`CIBW_BEFORE_BUILD` is very useful, absolutely key for projects that cannot satisfy everything with PIP dependencies, I just don't like showing or recommending `pip install .` (at least without serious warnings), and this is AFAICT the only reason you'd ever have that.

It would be much better to list your `setup_requires` contents in `CIBW_BEFORE_BUILD`, as that doesn't build your wheels twice.

Reply:
OK, then I completely agree!

Reply:
But I also agree with @joerick's point that in this overview, "Execute a shell command preparing each wheel's build if pyproject.toml is not enough" might be over the top? There's no mention of `install .` here either, so it's mostly when mentioning the examples that need a "are you sure?" clause about `pip install .`. No reason to bother the reader here, in the general overview of options? The way it sounds here is as if `CIBW_BEFORE_BUILD` is never a great idea.

Reply:
I'm quite okay to not add the extra comment here. I was just explaining my reasoning. I'm more worried about showing an example with `pip install .` in it later.

Reply:
OK, sorry for misunderstanding, then :-)

Reply:
Yes, this comment wasn't intended to refer to the entire PR, just this one line. Unfortunately it came first so it looked like that. Sorry if I was a little abrupt!

Reply:
```suggestion
|   | [`CIBW_BEFORE_BUILD`](https://cibuildwheel.readthedocs.io/en/stable/options/#before-build)  | Execute a shell command preparing each wheel's build |
```